crates/common/src/pbs/event.rs

```diff
@@ -0,0 +1,121 @@
+use std::net::SocketAddr;
+
+use alloy::{primitives::B256, rpc::types::beacon::relay::ValidatorRegistration};
+use axum::{
+    async_trait,
+    extract::State,
+    response::{IntoResponse, Response},
+    routing::post,
+    Json,
+};
+use eyre::bail;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
+use tokio::net::TcpListener;
+use tracing::{error, info, trace};
+
+use super::{
+    GetHeaderParams, GetHeaderReponse, SignedBlindedBeaconBlock, SubmitBlindedBlockResponse,
+};
+use crate::{
+    config::{load_env_var, BUILDER_SERVER_ENV},
+    pbs::BUILDER_EVENTS_PATH,
+};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum BuilderEvent {
+    GetHeaderRequest(GetHeaderParams),
+    GetHeaderResponse(Box<Option<GetHeaderReponse>>),
+    GetStatusEvent,
+    GetStatusResponse,
+    SubmitBlockRequest(Box<SignedBlindedBeaconBlock>),
+    SubmitBlockResponse(Box<SubmitBlindedBlockResponse>),
+    MissedPayload { block_hash: B256, relays: String },
+    RegisterValidatorRequest(Vec<ValidatorRegistration>),
+    RegisterValidatorResponse,
+}
+
+#[derive(Debug, Clone)]
+pub struct BuilderEventPublisher {
+    client: reqwest::Client,
+    endpoints: Vec<String>,
+}
+
+impl BuilderEventPublisher {
+    pub fn new(endpoints: Vec<String>) -> Self {
+        Self { client: reqwest::Client::new(), endpoints }
+    }
+
+    pub fn new_from_env() -> Option<Self> {
+        load_env_var(BUILDER_SERVER_ENV).map(|joined| {
+            let endpoints =
+                joined.split(',').map(|s| format!("http://{}{}", s, BUILDER_EVENTS_PATH)).collect();
+
+            Self::new(endpoints)
+        })
+    }
+
+    pub fn publish(&self, event: BuilderEvent) {
+        for endpoint in self.endpoints.clone() {
+            let client = self.client.clone();
+            let event = event.clone();
+
+            tokio::spawn(async move {
+                trace!("Publishing event to {}", endpoint);
```

Comment:
publishing refers more to a pubsub scenario, here seems more "sending event to blah"